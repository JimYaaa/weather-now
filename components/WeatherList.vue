<script lang="ts" setup>
import type { WeatherInfo } from '~~/composables/useWeather'
import type { Gecoding } from '~~/composables/useLocation'
import { useStorage } from '@vueuse/core'


const gecoding = defineModel('gecoding', {
    type: Object as PropType<Gecoding>,
})
const weatherStore = defineModel('weatherStore', {
    type: Array<any>,
    default: []
})
const props = defineProps({
    weatherInfo: {
        required: true,
        type: Object as PropType<WeatherInfo>
    }
})

const selectedWeatherIndex = defineModel('selectedWeatherIndex', {
    type: Number || null,
    default: null
})


useStorage('weatherStore', weatherStore)

function selectWeather(weatherInfo: WeatherInfo, index: number) {
    gecoding.value = weatherInfo.location
    selectedWeatherIndex.value = index
}

function addWeather(weatherInfo: WeatherInfo) {
    const { location, currentWeather, weatherForecast } = weatherInfo

    if (!currentWeather) return

    const isExited = weatherStore.value.find((weather: any) => (weather.id === location.id) && location.id)

    if (isExited) return

    weatherStore.value.push({
        id: location.id,
        name: location.name,
        country: location.country,
        currentWeather,
        weatherForecast,
        location,
    })

    selectedWeatherIndex.value = weatherStore.value.length - 1
    useLocalStorage('weatherStore', weatherStore.value)
}
</script>

<template>
    <div class="mt-8">
        <div class="flex justify-between items-center">
            <p class="text-5 font-bold font-sans">Weather</p>

            <button class="text-5" @click="addWeather(props.weatherInfo)">+</button>
        </div>

        <div class="w-full h-1px my-4 bg-white"></div>

        <ClientOnly fallback-tag="ul">
            <ul>
                <li
                    v-for="weather, index in weatherStore"
                    :key="weather.id"
                    :class="{ 'border-yellow': weather.id === weatherStore[selectedWeatherIndex]?.id }"
                    class="
                        flex justify-between items-stretch
                        p-4 mb-4 last:mb-0
                        border-solid border-1px border-white rounded-15px
                        cursor-pointer
                        hover:border-yellow
                        transition-all transition-500
                    "
                    @click="selectWeather(weather, index)"
                >
                    <div class="flex flex-col items-stretch justify-between">
                        <div>
                            <p class="font-bold">{{ weather.name }} {{ weather.name !== weather.country ? `/ ${weather.country}` : '' }}</p>
                        </div>
                    </div>
                </li>
            </ul>

            <template #fallback>
                <Icon class="color-white" name="svg-spinners:eclipse-half" size="50" />
            </template>
        </ClientOnly>
    </div>
</template>