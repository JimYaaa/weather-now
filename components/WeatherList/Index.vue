<script lang="ts" setup>
import type { WeatherInfo, WeatherStore } from '~~/entities/weather.ts'
import type { Gecoding } from '~~/entities/gecoding'
import { useStorage } from '@vueuse/core'


const gecoding = defineModel('gecoding', {
    type: Object as PropType<Gecoding>,
})

const weatherStorage = defineModel('weatherStorage', {
    type: Array<WeatherStore>,
    default: []
})

useStorage('weatherStorage', weatherStorage)

const isWidgetOpen = defineModel('isWidgetOpen', {
    type: Boolean,
    default: false
})

const props = defineProps({
    weatherInfo: {
        required: true,
        type: Object as PropType<WeatherInfo>
    }
})

const selectedWeatherIndex = defineModel('selectedWeatherIndex', {
    type: Number || null,
    default: null
})

function selectWeather(weatherInfo: WeatherInfo, index: number) {
    gecoding.value = weatherInfo.location
    selectedWeatherIndex.value = index
}

function addWeather(weatherInfo: WeatherInfo) {
    const { location, currentWeather, weatherForecast } = weatherInfo

    if (!currentWeather) return

   const isExited = weatherStorage.value.find((weather: WeatherStore) => (weather.id === location.id) && location.id)

   if (isExited) return

    weatherStorage.value.push({
        id: location.id,
        name: location.name,
        country: location.country,
        currentWeather,
        weatherForecast,
        location,
    })

    selectedWeatherIndex.value = weatherStorage.value.length - 1
}
</script>

<template>
    <div class="flex items-center flex-col flex-1 mt-8">
        <div class="w-full flex justify-between items-center">
            <p class="test text-5 font-bold font-sans">Weather</p>

            <button class="text-5" @click="addWeather(props.weatherInfo)">+</button>
        </div>

        <div class="w-full h-1px my-4 bg-white"></div>

        <div class="relative w-full flex-1">
            <ul class="absolute top-0 left-0 md-relative w-full h-full flex-1 md-max-h-400px overflow-auto">
                <li
                    v-for="weather, index in weatherStorage"
                    :key="weather.id"
                    :class="{ 'border-yellow': weather.id === weatherStorage[selectedWeatherIndex]?.id }"
                    class="
                        flex justify-between items-stretch
                        p-4 mb-4 last:mb-0
                        border-solid border-1px border-white rounded-15px
                        cursor-pointer
                        hover:border-yellow
                        transition-all transition-500
                    "
                    @click="selectWeather(weather, index)"
                >
                    <div class="flex flex-col items-stretch justify-between">
                        <div>
                            <p
                                data-test-id='weather-list-title'
                                class="sm-text-3 font-bold"
                            >
                                {{ weather.name }} {{ weather.name !== weather.country ? `/ ${weather.country}` : '' }}
                            </p>
                        </div>
                    </div>
                </li>
            </ul>
        </div>
        
        <div 
            class="
                flex justify-center items-center
                p-2 mt-4
                border-1px border-white border-solid rounded-50%
                md-hidden cursor-pointer
            "
            @click="() => isWidgetOpen = false"
        >
            <Icon
                class="color-white"
                name="ic:baseline-clear"
                size="35"
            />
        </div>
    </div>
</template>