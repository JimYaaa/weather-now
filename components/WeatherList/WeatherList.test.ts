import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, it, expect, afterEach, beforeEach } from 'vitest'
import WeatherList from './Index.vue'

describe('WeatherList', () => {
    const weatherInfo = {
        location: {
            id: 123,
            name: 'Tainan',
            latitude: 12,
            longitude: 45,
            country: 'Taiwan'
        },
        currentWeather: null,
        weatherForecast: [
            {
                temperature_celsius_max: 30,
                temperature_celsius_min: 28,
                temperature_celsius_unit: '°C',
                temperature_fahrenheit_max: 89,
                temperature_fahrenheit_min: 60,
                temperature_fahrenheit_unit: '°F',
                time: 'Thu',
                weather: 'Sunny'                
            }
        ]
    }

    let component: any

    beforeEach(async () => {
        component = await mountSuspended(WeatherList, {
            weatherInfo,
            weatherStorage: []
        })
    })

    afterEach(async () => {
        if (component) {
            await component.unmount()
        }
    })

    it('when location name equal location country only show one of them', async () => {
        await component.setProps({
            weatherInfo,
            weatherStorage: [
                {
                    "id": 1668285,
                    "name": "Taiwan",
                    "country": "Taiwan",             
                }
            ]
        })

        const listItem = await component.find('[data-test-id="weather-list-title"]')
        expect(listItem.text()).toBe('Taiwan')
        expect(component.html()).toMatchSnapshot()
    })

    it('when location name and country is different show both of them', async () => {
        await component.setProps({
            weatherInfo,
            weatherStorage: [
                {
                    "id": 1668284,
                    "name": "Osaka",
                    "country": "Japan",             
                }
            ]
        })

        const listItem = await component.find('[data-test-id="weather-list-title"]')
        expect(listItem.text()).toBe('Osaka / Japan')
        expect(component.html()).toMatchSnapshot()
    })
})