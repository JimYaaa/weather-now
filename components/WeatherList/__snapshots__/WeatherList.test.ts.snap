// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WeatherList > when location name and country is different show both of them 1`] = `
"<div class="mt-8">
  <div class="flex justify-between items-center">
    <p class="test text-5 font-bold font-sans">Weather</p><button class="text-5">+</button>
  </div>
  <div class="w-full h-1px my-4 bg-white"></div>
  <ul class="max-h-400px">
    <li class="flex justify-between items-stretch p-4 mb-4 last:mb-0 border-solid border-1px border-white rounded-15px cursor-pointer hover:border-yellow transition-all transition-500">
      <div class="flex flex-col items-stretch justify-between">
        <div>
          <p data-test-id="weather-list-title" class="sm-text-3 font-bold">Osaka / Japan</p>
        </div>
      </div>
    </li>
  </ul>
</div>"
`;

exports[`WeatherList > when location name equal location country only show one of them 1`] = `
"<div class="mt-8">
  <div class="flex justify-between items-center">
    <p class="test text-5 font-bold font-sans">Weather</p><button class="text-5">+</button>
  </div>
  <div class="w-full h-1px my-4 bg-white"></div>
  <ul class="max-h-400px">
    <li class="flex justify-between items-stretch p-4 mb-4 last:mb-0 border-solid border-1px border-white rounded-15px cursor-pointer hover:border-yellow transition-all transition-500">
      <div class="flex flex-col items-stretch justify-between">
        <div>
          <p data-test-id="weather-list-title" class="sm-text-3 font-bold">Taiwan </p>
        </div>
      </div>
    </li>
  </ul>
</div>"
`;
